File: Gramaticas/Ejercicio2/C/ejercicio2.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

// L(G2) = { a^n b^(n+1) | n >= 0 }
bool es_ejercicio2(char *s) {
    int i = 0;
    int a = 0;
    int b = 0;

    // Contar 'a' del inicio
    while (s[i] == 'a') {
        a++;
        i++;
    }

    // Contar 'b' que siguen
    while (s[i] == 'b') {
        b++;
        i++;
    }

    // Si hay otros símbolos, no pertenece
    if (s[i] != '\0') {
        return false;
    }

    // Regla: b = a + 1
    return (b == a + 1);
}

int main() {
    FILE *f = fopen("ejercicio2.txt", "r");
    if (!f) {
        printf("Error al abrir g2.txt\n");
        return 1;
    }

    char linea[256];
    while (fgets(linea, sizeof(linea), f)) {
        // quitar salto de línea
        linea[strcspn(linea, "\n")] = 0;

        if (es_ejercicio2(linea)) {
            printf("%s => ACEPTA\n", linea);
        } else {
            printf("%s => NO ACEPTA\n", linea);
        }
    }

    fclose(f);
    return 0;
}

File: Gramaticas/Ejercicio2/Python/ejercicio2.py
# L(G2) = { a^n b^(n+1) | n >= 0 }
# Idea: contar cuántas 'a' hay al inicio, luego cuántas 'b' siguen,
# y verificar que no haya otros símbolos y que b = a + 1.

def es_ejercicio2(cadena):
    i = 0
    a_count = 0

    # Contar 'a' iniciales
    while i < len(cadena) and cadena[i] == "a":
        a_count += 1
        i += 1

    # Contar 'b' después de las 'a'
    b_count = 0
    while i < len(cadena) and cadena[i] == "b":
        b_count += 1
        i += 1

    # Si quedó algún otro símbolo, no pertenece
    if i != len(cadena):
        return False

    # Regla: debe haber exactamente una 'b' más que 'a'
    return b_count == a_count + 1

def procesar(archivo):
    with open(archivo, "r") as f:
        for linea in f:
            cad = linea.strip()
            if es_ejercicio2(cad):
                print(cad, "=> ACEPTA")
            else:
                print(cad, "=> NO ACEPTA")

if __name__ == "__main__":
    procesar("ejercicio2.txt")
